plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/ChangeVisibilityFixFactories.kt

```diff
@@ -0,0 +1,69 @@
+import org.jetbrains.kotlin.idea.KotlinBundle
+import org.jetbrains.kotlin.idea.api.applicator.HLApplicatorInput
+import org.jetbrains.kotlin.idea.api.applicator.applicator
+import org.jetbrains.kotlin.idea.fir.api.fixes.HLApplicatorTargetWithInput
+import org.jetbrains.kotlin.idea.fir.api.fixes.diagnosticFixFactory
+import org.jetbrains.kotlin.idea.fir.api.fixes.withInput
+import org.jetbrains.kotlin.idea.util.isRedundantSetter
+import org.jetbrains.kotlin.idea.util.removeRedundantSetter
+import org.jetbrains.kotlin.lexer.KtModifierKeywordToken
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.visibilityModifierType
+
+object ChangeVisibilityFixFactories {
+
+    class Input(val elementName: String) : HLApplicatorInput
+
+    private fun getApplicator(visibilityModifier: KtModifierKeywordToken, forceUsingExplicitModifier: Boolean) =
+        applicator<KtModifierListOwner, Input> {
+            if (forceUsingExplicitModifier) {
+                familyName { KotlinBundle.message("make.0.explicitly", visibilityModifier) }
+                actionName { psi, input -> KotlinBundle.message("make.0.1.explicitly", input.elementName, visibilityModifier) }
+            } else {
+                familyName { KotlinBundle.message("make.0", visibilityModifier) }
+                actionName { psi, input -> KotlinBundle.message("make.0.1", input.elementName, visibilityModifier) }
+            }
+            applyTo { psi, input ->
+                // TODO: also add logic to change visibility of expect/actual declarations.
+                if (visibilityModifier == KtTokens.PUBLIC_KEYWORD && !forceUsingExplicitModifier) {
+                    psi.visibilityModifierType()?.let { psi.removeModifier(it) }
+                } else {
+                    psi.addModifier(visibilityModifier)
+                }
+
+                val propertyAccessor = psi as? KtPropertyAccessor
+                if (propertyAccessor?.isRedundantSetter() == true) {
+                    removeRedundantSetter(propertyAccessor)
+                }
+            }
+        }
+
+    val makePublicExplicitApplicator = getApplicator(KtTokens.PUBLIC_KEYWORD, true)
+
+    val noExplicitVisibilityInApiMode =
+        diagnosticFixFactory(KtFirDiagnostic.NoExplicitVisibilityInApiMode::class, makePublicExplicitApplicator) { diagnostic ->
+            createFixForNoExplicitVisibilityInApiMode(diagnostic.psi)
+        }
+    val noExplicitVisibilityInApiModeWarning =
+        diagnosticFixFactory(KtFirDiagnostic.NoExplicitVisibilityInApiModeWarning::class, makePublicExplicitApplicator) { diagnostic ->
+            createFixForNoExplicitVisibilityInApiMode(diagnostic.psi)
+        }
+
+    private fun createFixForNoExplicitVisibilityInApiMode(declaration: KtDeclaration): List<HLApplicatorTargetWithInput<KtDeclaration, Input>> {
+        val name = when (declaration) {
+            is KtConstructor<*> -> "<init>"
```

Comment:
We don't have a centralized place for this special name?!

Reply:
Good point! Actually we do in `SpecialNames`. Let me use that instead.